File: pie_chart.py
import math
from dataclasses import dataclass
from language_colour_map import get_language_colour
from create_svg import create_svg

@dataclass
class Pos:
    x: int
    y: int

    def to_cartesian(self, radius: int, angle_in_degrees: int) -> 'Pos':
        angle_in_radians = (angle_in_degrees - 90) * math.pi / 180
        x = self.x + (radius * math.cos(angle_in_radians))
        y = self.y + (radius * math.sin(angle_in_radians))

        return Pos(x, y)

@dataclass
class Arc:
    radius: int
    start: Pos
    end: Pos
    laf: '0' or '1'

def create_arc_data(pos: Pos, radius: int, start_angle: int, end_angle: int) -> Arc:
    start = pos.to_cartesian(radius, end_angle)
    end = pos.to_cartesian(radius, start_angle)
    laf = '0' if end_angle - start_angle <= 180 else '1'

    return Arc(radius, start, end, laf)

def create_arc(center: Pos, arc_data: Arc) -> str:
    radius = arc_data.radius
    start_x = math.ceil(arc_data.start.x)
    start_y = math.ceil(arc_data.start.y)
    end_x = math.ceil(arc_data.end.x)
    end_y = math.ceil(arc_data.end.y)
    laf = arc_data.laf

    return f"M {center.x} {center.y} L {start_x} {start_y} A {radius} {radius} 0 {laf} 0 {end_x} {end_y} L {center.x} {center.y}"

def create_pie_chart(svg_size: int, lang_map: dict[str, str]) -> str:
    total = sum([v for v in lang_map.values()])
    center = Pos(svg_size // 2, svg_size // 2)
    radius = svg_size // 2
    data = []
    cumulative_total = 0
    for lang, count in lang_map.items():
        angle_coverage = round((count / total) * 360)
        arc_data = create_arc_data(center, radius, cumulative_total, cumulative_total+angle_coverage)
        colour = get_language_colour(lang)

        data.append(f'<path class="{lang}" d="{create_arc(center, arc_data)}" fill="{colour}" stroke="{colour}"><title>{lang}</title></path>')
        cumulative_total += angle_coverage

    create_svg(svg_size, svg_size, '\n'.join(data), 'pie_chart')
    return '![Pie Chart](./assets/pie_chart.svg "Pie Chart detailing languages used")'

File: src/pieChart.nim
import
  std/[
    math,
    strformat,
    tables,
    enumerate
  ],
  language, colours, svg, class

class Pos:
  x: int
  y: int

  func newPos(x, y: int): Pos {.static.} =
    Pos(x: x, y: y)
  
  func toCartesian(self: Pos, radius, angleInDegrees: int): Pos =
    new result
    let 
      angleInRadians = (angleInDegrees - 90).toFloat * PI / 180.toFloat
      x = (self.x.toFloat + (radius.toFloat * cos(angleInRadians))).floor.toInt
      y = (self.y.toFloat + (radius.toFloat * sin(angleInRadians))).floor.toInt
    
    newPos(x, y)

type LargeArcFlag = enum
    Zero, One

func `$`(largeArcFlag: LargeArcFlag): string =
  case largeArcFlag:
  of LargeArcFlag.Zero: "0"
  of LargeArcFlag.One: "1"

class Arc:
  radius: int
  startPos: Pos
  endPos: Pos
  largeArcFlag: LargeArcFlag


  func newArc(pos: Pos, radius, startAngle, endAngle: int): Arc {.static.} =
    new result
    result.radius = radius
    result.startPos = pos.toCartesian(radius, endAngle)
    result.endPos = pos.toCartesian(radius, startAngle)
    result.largeArcFlag = if endAngle - startAngle <= 180: LargeArcFlag.Zero else: LargeArcFlag.One

func createArc(center: Pos, arc: Arc): string =
  fmt"M {center.x} {center.y} L {arc.startPos.x} {arc.startPos.y} A {arc.radius} {arc.radius} 0 {$arc.largeArcFlag} 0 {arc.endPos.x} {arc.endPos.y} L {center.x} {center.y}"

proc createPieChart*(svgSize: int, map: LanguageMap): string =
  let 
    total  = map.total
    center = newPos((svgSize / 2).toInt, (svgSize / 2).toInt)
    radius = (svgSize / 2).toInt

  var
    cumulativeTotal = 0
    svg = newSvg(svgSize, svgSize)

  for index, (language, count) in enumerate(map.pairs):
    let 
      angleCoverage = ((count / total) * 360).toInt
      arc = newArc(center, radius, cumulativeTotal, cumulativeTotal + angleCoverage)
      colour = getLanguageColour(language)
      arcSvg = newSvgNode(Path,children = @[newSvgNode(Title, children = @[newSvgNode(Text, inner = language)])],attrs = @[newSvgAttribute("class", language),newSvgAttribute("d", createArc(center, arc)),newSvgAttribute("fill", colour),newSvgAttribute("stroke", colour)])
    
    svg.add(arcSvg)

    cumulativeTotal += angleCoverage
  
  svg.write("PieChart")
  "![Pie Chart](./assets/PieChart.svg \"Pie Chart Detailing used languages\")"